### PR TITLE
Migrate to EndpointSlices API 

### PR DESCRIPTION
**Pull request by jnoordsij**
### What does this PR do?

Switch to Kubernetes EndpointSlices API as described in #10638.

### Motivation

Discovered existence of this API while looking into behavior regarding non-ready service endpoints still being served traffic. Although this change will probably make little changes in that regard, it's probably still a technical improvement that allows for more features in the future and still might slightly improve performance.

### More

- [x] Added/updated tests
- [ ] Added/updated documentation

### To Do

- [ ] Create a list of important changes
- [ ] Update documentation; possibly with some draft for upgrade notes
- [ ] Check for additionally useful tests

### CODE REVIEW

**Comment by jnoordsij** [comment]
Any early comments or suggestions are welcome, but given some tests are still failing and I'm aiming to write down a more descriptive summary and list of important changes, a thorough review is probably something to only do at a later point. 

I'll hope to continue working on this soon. Yet like I said, any comments, concerns or suggestions are welcome in the meantime!

**Comment by nmengin** [comment]
Hello @jnoordsij,

Thank you for the answer.
We are waiting for your go before starting the review 😄

**Comment by jnoordsij** [comment]
@nmengin I think this is all ready for a first review right now!

File: pkg/provider/kubernetes/ingress/kubernetes.go [hunk 1]
```diff
@@ -668,12 +669,21 @@ func (p *Provider) loadService(client Client, namespace string, backend netv1.In
 
 		protocol := getProtocol(portSpec, portName, svcConfig)
 
-		for _, addr := range subset.Addresses {
-			hostPort := net.JoinHostPort(addr.IP, strconv.Itoa(int(port)))
+		for _, endpoint := range endpointSlice.Endpoints {
+			if !(*endpoint.Conditions.Ready) {
+				continue
+			}
```
Comment (rtribotte) [inline pkg/provider/kubernetes/ingress/kubernetes.go:675]: Hello @jnoordsij!

Thanks for this contribution!

I'm wondering if filtering the endpoint condition here is introducing a breaking change in terms of discovery, WDYT?

Comment (jnoordsij) [reply pkg/provider/kubernetes/ingress/kubernetes.go:675]: I did think at first when starting with this there might be a difference, but from what I've gathered it should not change anything. The endpoints with `ready=true` state on the `EndpointSlice` API should match the `addresses` on the `Endpoints` API (which also has a subsequent `notReadyAddresses` entry) and thus should return the exact same set of endpoints as before.

See also the [following note from documentation](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#serving):
> Although serving is almost identical to ready, it was added to prevent breaking the existing meaning of ready. It may be unexpected for existing clients if ready could be true for terminating endpoints, since historically terminating endpoints were never included in the Endpoints or EndpointSlice API to begin with. For this reason, ready is always false for terminating endpoints, and a new condition serving was added in v1.20 so that clients can track readiness for terminating pods independent of the existing semantics for ready.

Basically the new API has more information about state: rather than only doing a 'binary' separation between `ready` and not `ready`, there are now three conditions you can check on depending on your preference: `ready`, `serving` and `terminating`.

Comment (rtribotte) [reply pkg/provider/kubernetes/ingress/kubernetes.go:675]: Indeed, my first assumption was that states were not coexisting but rather a value for the state of the endpoint. Thanks for putting some thoughts and providing clarification on this 👍

**Comment by kevinpollet** [comment]
Hello @jnoordsij and thanks for your contribution,

To support Gateway API v1.1.0, the use of the EndpointSlices API is mandatory.

Therefore, we would like to move quickly on this pull request. Can we rebase and push some changes to it?

**Comment by jnoordsij** [comment]
@kevinpollet sure thing! Feel free to further alter it where needed and if you have any questions regarding the work feel free to ask.

**Review by kevinpollet — APPROVED** [review]
Thanks 👍

**Review by rtribotte — APPROVED** [review]
Thanks 👍

File: docs/content/migration/v3.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+description: "Learn the steps needed to migrate to new Traefik Proxy v3 versions. Read the technical documentation."
+---
+
+# Migration: Steps needed between the versions
+
+## v3.0 to v3.1
+
+### Kubernetes Provider RBACs
+
+Starting with v3.1, the Kubernetes Providers now use the [EndpointSlices API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/) to discover service endpoint addresses.
```
Comment (rtribotte) [inline docs/content/migration/v3.md:12]:
```suggestion
Starting with v3.1, the Kubernetes Providers now use the [EndpointSlices API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/) (Kubernetes >=v1.21) to discover service endpoint addresses.
```

**Review by mmatur — APPROVED** [review]
